src/aspire/source/micrograph.py

```diff
@@ -1,75 +1,388 @@
 import logging
+import os
+from abc import ABC, abstractmethod
+from glob import glob
+from pathlib import Path
 
 import numpy as np
 
 from aspire.image import Image
 from aspire.source import Simulation
 from aspire.source.image import _ImageAccessor
+from aspire.storage import StarFile
 from aspire.utils import Random, grid_2d
+from aspire.volume import Volume
 
 logger = logging.getLogger(__name__)
 
 
-class MicrographSimulation:
+class MicrographSource(ABC):
+    def __init__(self, micrograph_count, micrograph_size, dtype):
+        """ """
+        self.micrograph_count = int(micrograph_count)
+        self.micrograph_size = int(micrograph_size)
+        self.dtype = np.dtype(dtype)
+
+        self._images_accessor = _ImageAccessor(self._images, self.micrograph_count)
+
+    def __repr__(self):
+        """
+        String representation.
+
+        :return: Returns a string description of instance.
+        """
+        return f"{self.__class__.__name__} with {self.micrograph_count} {self.dtype.name} micrographs of size {self.micrograph_size}x{self.micrograph_size}"
+
+    def __len__(self):
+        """
+        Returns the number of micrographs.
+
+        :return: Returns the micrograph count
+        """
+        return self.micrograph_count
+
+    def save(self, path, name_prefix="micrograph", overwrite=True):
+        """
+        Save micrographs to `path`.
+
+        `path` should be a directory. If it does not exist, ASPIRE will attempt to create it.
+
+        Currently saves micrographs to `.mrc`.
+
+        :param path: Directory to save data.
+        :param name_prefix: Optional, name prefix string for micrograph files.
+        :param overwrite: Optional, bool. Allow writing to existing directory,
+            and overwriting existing files.
+        :return: List of saved `.mrc` files.
+        """
+
+        # Make dir if does not exist.
+        Path(path).mkdir(parents=True, exist_ok=overwrite)
```

Comment:
No. If they wish not to overwrite things they should not enable overwrite. They should move the folder or micrographs if that's a problem.

Lets not be silly.

Reply:
To be clearer, what this saves is essentially a directories worth of stuff.  This stuff contains both MRC and STAR files.  I would take the view that this directory (in its entirety) is essentially the "thing" that is being saved.


```
mg_sim1
├── micrograph_0.mrc
├── micrograph_0.star
├── micrograph_1.mrc
├── micrograph_1.star
├── micrograph_2.mrc
└── micrograph_2.star
mg_sim2
├── micrograph_0.mrc
├── micrograph_0.star
├── micrograph_1.mrc
├── micrograph_1.star
├── micrograph_2.mrc
└── micrograph_2.star
```

It is possible to reuse a dir, or use the cwd, and that is allowed via `overwrite`. I dont think that is the best way to use the method.

Our `StarFile.write` has no notion of `overwrite`. It will just plow over stuff. So that would need to be fixed, and tested, etc or we implement logic that belongs to `StarFile` in `MicrographSource` (yugg).

Overall I would say that ASPIRE should not be a replacement for basic commandline skills like moving directories etc.

If this feature is confusing we can make the default to be `False` and force a user to always have a fresh directory unless they explicitly override.  I just assumed `overwrite=True` was a more useful (less likely to crash) default, until we address #892 and #745 . If those issues are more important now, we can move them up the priority list.

Reply:
Thanks for the explanation! That makes sense.